Begin drawing pixels to the screen from site isolated iframes

achristensen07 opened the pull request:
#### 90549aec61bacbaee7383e1fb2623cf34edc9273
<pre>
Begin drawing pixels to the screen from site isolated iframes
<a href="https://bugs.webkit.org/show_bug.cgi?id=251245">https://bugs.webkit.org/show_bug.cgi?id=251245</a>
rdar://104726144

Reviewed by Simon Fraser.

If we have a remote frame, the parent process needs to make a layer but put nothing in it.
The iframe process&apos;s root layer needs to be the contents of the empty layer.
To make this happen, the journey starts in RenderLayerBacking::updateConfiguration where
we call GraphicsLayerCA::setContentsToPlatformLayerHost, which makes this layer a host for
content that will come from another process with a given identifier generated in the UI
process.  We need the iframe process to know it has a hosting context identifier
so we do this by passing the LayerHostingContextIdentifier to the WebPage.  If this is present,
then in RemoteLayerTreeContext::buildTransaction we set the layerHostingContextIdentifier on the
RemoteLayerTreeTransaction.  When the UI process sees a transaction with a layerHostingContextIdentifier
it has some special logic in RemoteLayerTreeDrawingAreaProxy::commitLayerTree to make its
&quot;root layer&quot; not a root layer of the whole page, but rather it sets its &quot;root layer&quot; to be a
sublayer of the parent process&apos;s empty layer which was intended for this purpose.

* LayoutTests/http/tests/site-isolation/basic-iframe-expected.html: Added.
* LayoutTests/http/tests/site-isolation/basic-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/resources/green-background.html: Added.
* Source/WebCore/html/HTMLFrameOwnerElement.cpp:
(WebCore::HTMLFrameOwnerElement::disconnectContentFrame):
* Source/WebCore/html/HTMLFrameOwnerElement.h:
* Source/WebCore/page/RemoteFrameView.h:
(isType):
* Source/WebCore/platform/Widget.h:
(WebCore::Widget::isRemoteFrameView const):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::setRemoteFrame):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebCore/rendering/RenderWidget.cpp:
(WebCore::RenderWidget::requiresAcceleratedCompositing const):
(WebCore::RenderWidget::remoteFrame const):
* Source/WebCore/rendering/RenderWidget.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
(WebKit::RemoteLayerTreeTransaction::setRemoteFrameIdentifier):
(WebKit::RemoteLayerTreeTransaction::remoteFrameIdentifier const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::encode const):
(WebKit::RemoteLayerTreeTransaction::decode):
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::ProvisionalFrameProxy):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::RemoteLayerTreeDrawingAreaProxy::frameLayerMap):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::layerWillBeUsedForRemoteFrame):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateLayerTree):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::parentFrame):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
(WebKit::GraphicsLayerCARemote::setRemoteFrame):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::buildTransaction):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::didCommitLoadInAnotherProcess):
Add a call to scheduleInvalidateStyleAndLayerComposition or the logic in
RenderWidget::requiresAcceleratedCompositing won&apos;t be called until something else
invalidates style.  I found this initially by being frustrated that
requiresAcceleratedCompositing returning true didn&apos;t do anything, then noticing that
when I plugged my computer into power it produced a layer.  This call does a similar
kick of the iframe to tell it to make a layer now, when the LocalFrame is swapped
out with a RemoteFrame.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::parentFrameIdentifier const):

Canonical link: <a href="https://commits.webkit.org/259937@main">https://commits.webkit.org/259937@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1c37d8eeba961b141d80d10c380ad8e380c2567b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/106467 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/15489 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/39281 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/115653 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/175752 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/110374 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/16979 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/6714 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/98664 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/115316 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/112234 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/12896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/95870 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/40456 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/109950 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/94771 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/27514 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/82169 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/8721 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/28866 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/9255 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/5440 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/14875 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/48412 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6865 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/10802 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/GraphicsLayer.h:
```diff
@@ -530,6 +531,8 @@ class GraphicsLayer : public RefCounted<GraphicsLayer> {
 #endif
     virtual bool usesContentsLayer() const { return false; }
 
+    virtual void setRemoteFrame(RemoteFrame&) { }
```
This is a layering violation. RemoteFrame is a `page` concept, and here we're in `platform` code.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
```diff
@@ -330,6 +334,8 @@ class RemoteLayerTreeTransaction {
     HashSet<RefPtr<PlatformCALayerRemote>> m_changedLayers; // Only used in the Web process.
     LayerPropertiesMap m_changedLayerProperties; // Only used in the UI process.
 
+    Markable<WebCore::FrameIdentifier> m_remoteFrameIdentifier;
```
It's not super clear what this is. Is it the identifier for the rootmost frame in the process that's generating this transaction? We need a name for that concept (rootmost frame?)

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
```diff
@@ -67,6 +67,8 @@ class RemoteLayerTreeDrawingAreaProxy : public DrawingAreaProxy {
     void updateOverlayRegionIDs(const HashSet<WebCore::GraphicsLayer::PlatformLayerID> &overlayRegionIDs) { m_remoteLayerTreeHost->updateOverlayRegionIDs(overlayRegionIDs); }
 #endif
 
+    HashMap<WebCore::FrameIdentifier, WebCore::GraphicsLayer::PlatformLayerID>& frameLayerMap() { return m_frameLayerMap; }
```
This might be better in `RemoteLayerTreeHost` which has a number of such maps (but keyed by layer ID)

smfr replied:
I see that entries in this map are transient. It's bad form to have member variables on long-lived objects representing temporary state like this; this object should live on the stack, or go along with the transaction.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
```diff
@@ -111,6 +113,7 @@ class RemoteLayerTreeDrawingAreaProxy : public DrawingAreaProxy {
     virtual void didCommitLayerTree(const RemoteLayerTreeTransaction&, const RemoteScrollingCoordinatorTransaction&) { }
 
     void asyncSetLayerContents(WebCore::GraphicsLayer::PlatformLayerID, ImageBufferBackendHandle&&);
+    void layerWillBeUsedForRemoteFrame(WebCore::GraphicsLayer::PlatformLayerID, WebCore::FrameIdentifier);
```
I think we should talk about "hosting a frame" (like `ScrollingTreeFrameHostingNode`)

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -133,11 +133,14 @@
-
-    bool didUpdateEditorState = layerTreeTransaction.hasEditorState() && m_webPageProxy.updateEditorState(layerTreeTransaction.editorState(), WebPageProxy::ShouldMergeVisualEditorState::Yes);
+    bool didUpdateEditorState { false };
+    if (!layerTreeTransaction.remoteFrameIdentifier()) {
```
Maybe we want a totally separate entry point for transactions coming in for remote frames. Much of the code in this function assumes main frame.

achristensen07 replied:
There's enough in common I think this approach is better for now, but not by a lot.  That may not always be the case, though.  If you feel strongly about it I can separate it now or later.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -200,6 +203,11 @@
 
+void RemoteLayerTreeDrawingAreaProxy::layerWillBeUsedForRemoteFrame(WebCore::GraphicsLayer::PlatformLayerID layerID, WebCore::FrameIdentifier frameID)
+{
+    m_frameLayerMap.set(frameID, layerID);
```
We gotta remove them too somewhere.

smfr replied:
Oh, I see you `take()` them later. See my comment about member variables containing transient state.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
```diff
@@ -135,5 +136,9 @@ RefPtr<WebCore::GraphicsLayerAsyncContentsDisplayDelegate> GraphicsLayerCARemote
 
+void GraphicsLayerCARemote::setRemoteFrame(RemoteFrame& frame)
+{
+    WebProcess::singleton().parentProcessConnection()->send(Messages::RemoteLayerTreeDrawingAreaProxy::LayerWillBeUsedForRemoteFrame(primaryLayerID(), frame.frameID()), m_context->drawingAreaIdentifier());
```
I don't think this should IPC. It should just store some ID on the layer which gets to the UI process, where layers are glued together.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
```diff
@@ -151,6 +152,8 @@
 
     PlatformCALayerRemote& rootLayerRemote = downcast<PlatformCALayerRemote>(rootLayer);
     transaction.setRootLayerID(rootLayerRemote.layerID());
+    if (m_webPage.parentFrameIdentifier())
```
OK my head is starting to hurt with "parent frame", and "main frame"

achristensen07 replied:
This will be resolved by future frame tree model improvements.

smfr commented on Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
```diff
@@ -151,6 +152,8 @@
     PlatformCALayerRemote& rootLayerRemote = downcast<PlatformCALayerRemote>(rootLayer);
     transaction.setRootLayerID(rootLayerRemote.layerID());
+    if (m_webPage.parentFrameIdentifier())
+        transaction.setRemoteFrameIdentifier(m_webPage.mainWebFrame().frameID());
```
`rootmostFrame().frameID()` ?

achristensen07 replied:
This will be resolved by future frame tree model improvements.

smfr commented on Source/WebCore/html/HTMLFrameOwnerElement.cpp:
```diff
@@ -76,8 +78,15 @@ void HTMLFrameOwnerElement::clearContentFrame()
+    if (RefPtr frame = m_contentFrame.get()) {
+        switch (frame->frameType()) {
+        case AbstractFrame::FrameType::Local:
+            downcast<LocalFrame>(*frame).loader().frameDetached();
```
Why not virtual function on Frame?

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -375,6 +375,12 @@ Ref<PlatformCALayer> GraphicsLayerCA::createPlatformCALayer(Ref<WebCore::Model>,
 
+Ref<PlatformCALayer> GraphicsLayerCA::createPlatformCALayerHost(LayerHostingContextIdentifier, PlatformCALayerClient* owner)
+{
+    ASSERT_NOT_REACHED();
```
Maybe a comment to say why this isn't reached here.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
```diff
@@ -133,11 +133,14 @@
-
-    bool didUpdateEditorState = layerTreeTransaction.hasEditorState() && m_webPageProxy.updateEditorState(layerTreeTransaction.editorState(), WebPageProxy::ShouldMergeVisualEditorState::Yes);
+    bool didUpdateEditorState { false };
+    if (!layerTreeTransaction.remoteContextHostIdentifier()) {
```
Would prefer something more readable (and flipped), like `if (isLayerTreeTransaction.isMainFrameTransaction())` or something.

achristensen07 replied:
I'm going to use isMainFrameProcessTransaction because the transaction can include layers from many frames in one process.

achristensen07 committed:
Begin drawing pixels to the screen from site isolated iframes
https://bugs.webkit.org/show_bug.cgi?id=251245
rdar://104726144

Reviewed by Simon Fraser.

If we have a remote frame, the parent process needs to make a layer but put nothing in it.
The iframe process's root layer needs to be the contents of the empty layer.
To make this happen, the journey starts in RenderLayerBacking::updateConfiguration where
we call GraphicsLayerCA::setContentsToPlatformLayerHost, which makes this layer a host for
content that will come from another process with a given identifier generated in the UI
process.  We need the iframe process to know it has a hosting context identifier
so we do this by passing the LayerHostingContextIdentifier to the WebPage.  If this is present,
then in RemoteLayerTreeContext::buildTransaction we set the layerHostingContextIdentifier on the
RemoteLayerTreeTransaction.  When the UI process sees a transaction with a layerHostingContextIdentifier
it has some special logic in RemoteLayerTreeDrawingAreaProxy::commitLayerTree to make its
"root layer" not a root layer of the whole page, but rather it sets its "root layer" to be a
sublayer of the parent process's empty layer which was intended for this purpose.

* LayoutTests/http/tests/site-isolation/basic-iframe-expected.html: Added.
* LayoutTests/http/tests/site-isolation/basic-iframe.html: Added.
* LayoutTests/http/tests/site-isolation/resources/green-background.html: Added.
* Source/WebCore/html/HTMLFrameOwnerElement.cpp:
(WebCore::HTMLFrameOwnerElement::disconnectContentFrame):
* Source/WebCore/html/HTMLFrameOwnerElement.h:
* Source/WebCore/page/RemoteFrameView.h:
(isType):
* Source/WebCore/platform/Widget.h:
(WebCore::Widget::isRemoteFrameView const):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::setRemoteFrame):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateConfiguration):
* Source/WebCore/rendering/RenderWidget.cpp:
(WebCore::RenderWidget::requiresAcceleratedCompositing const):
(WebCore::RenderWidget::remoteFrame const):
* Source/WebCore/rendering/RenderWidget.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.h:
(WebKit::RemoteLayerTreeTransaction::setRemoteFrameIdentifier):
(WebKit::RemoteLayerTreeTransaction::remoteFrameIdentifier const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::encode const):
(WebKit::RemoteLayerTreeTransaction::decode):
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/ProvisionalFrameProxy.cpp:
(WebKit::ProvisionalFrameProxy::ProvisionalFrameProxy):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.h:
(WebKit::RemoteLayerTreeDrawingAreaProxy::frameLayerMap):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.messages.in:
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeDrawingAreaProxy.mm:
(WebKit::RemoteLayerTreeDrawingAreaProxy::commitLayerTree):
(WebKit::RemoteLayerTreeDrawingAreaProxy::layerWillBeUsedForRemoteFrame):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::updateLayerTree):
* Source/WebKit/UIProcess/WebFrameProxy.h:
(WebKit::WebFrameProxy::parentFrame):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.cpp:
(WebKit::GraphicsLayerCARemote::setRemoteFrame):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeContext.mm:
(WebKit::RemoteLayerTreeContext::buildTransaction):
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp:
(WebKit::WebFrame::didCommitLoadInAnotherProcess):
Add a call to scheduleInvalidateStyleAndLayerComposition or the logic in
RenderWidget::requiresAcceleratedCompositing won't be called until something else
invalidates style.  I found this initially by being frustrated that
requiresAcceleratedCompositing returning true didn't do anything, then noticing that
when I plugged my computer into power it produced a layer.  This call does a similar
kick of the iframe to tell it to make a layer now, when the LocalFrame is swapped
out with a RemoteFrame.
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
* Source/WebKit/WebProcess/WebPage/WebPage.h:
(WebKit::WebPage::parentFrameIdentifier const):

Canonical link: https://commits.webkit.org/259937@main
